perch-agent/src/agent/middleware/logging.py

```
@@ -0,0 +1,137 @@
+import logging
+import time
+from collections import deque
+from collections.abc import Awaitable, Callable
+from typing import Any
+
+# Cap the per-request tool-call history. The middleware instance is already
+# request-scoped (constructed inside stream_chat), so this just bounds memory
+# under a runaway agent loop — older entries are evicted before append.
+_TOOL_CALL_HISTORY_LIMIT = 200
+
+from langchain.agents.middleware import AgentMiddleware, ModelRequest, ModelResponse
```

Ruff (0.15.12)

[error] Module level import not at top of file

(E402)